File: antia_numerical_methods/f77/sub16/gausrc.f
!     To calculate weights and abscissas of a quadrature formula with
!     specified weight function when recurrence relation for orthogonal
!     polynomial is known.
!
!     N : (input) Number of points in the required quadrature formula
!     W : (output) Array of length N, which will contain the weights
!     AB : (output) Array of length N containing the abscissas
!     COF : (input) Array of length 3*N containing the coefficients of
!		the recurrence relation for orthogonal polynomials
!		P_i(x)=(COF(1,i)*x+COF(2,i))P_{i-1}(x) - COF(3,i)*P_{i-2}(x)
!     RI0 : (input) The integral of weight function over the required
!		interval.
!     IER : (output) Error parameter, IER=0 implies successful execution
!		IER=302 implies N.LE.0
!		IER=321 implies that some coefficient becomes imaginary
!			during calculations.
!		In both these cases calculations are abandoned.
!		Other values may be set by TQL2
!     WK : Real array of length N*(N+2) used as scratch space
!
!     Required routines : TQL2
 
      SUBROUTINE GAUSRC(N,W,AB,COF,RI0,IER,WK)
      IMPLICIT REAL*16(A-H,O-Z)
      PARAMETER (REPS=1.Q-33)
!      PARAMETER (REPS=1.D-15)
!	For REAL*4 use REPS=6.E-8
!      PARAMETER (REPS=6.E-8)
      DIMENSION WK(N,N+2),COF(3,N),W(N),AB(N)
 
      IER=302
      IF(N.LE.0) RETURN
      LJ=N
 
!     Calculate the coefficients of symmetric tridiagonal matrix
      DO 2000 I=1,N
        WK(I,N+1)=-COF(2,I)/COF(1,I)
        IF(I.LT.N) THEN
          R1=COF(3,I+1)/(COF(1,I)*COF(1,I+1))
          IF(R1.GE.0.0) THEN
            WK(I+1,N+2)=SQRT(R1)
          ELSE
            IER=321
            RETURN
          ENDIF
        ENDIF
        DO 1800 J=1,N
          WK(J,I)=0.0
1800    CONTINUE
        WK(I,I)=1.0
2000  CONTINUE
 
!     Find eigenvalues and eigenvectors of the tridiagonal matrix
      CALL TQL2(WK,N,LJ,WK(1,N+1),WK(1,N+2),REPS,IER)
      IF(IER.GT.0) RETURN
 
!     Calculate the abscissas and weights
      DO 3000 I=1,N
        AB(I)=WK(I,N+1)
        W(I)=WK(1,I)**2*RI0
3000  CONTINUE
 
      END

File: projects/antia_numerical_methods@Beliavsky/c/fun/gausrc.c
/*	To calculate weights and abscissas of a quadrature formula with
	specified weight function when recurrence relation for orthogonal
	polynomial is known.

	N : (input) Number of points in the required quadrature formula
	W : (output) Array of length N, which will contain the weights
	AB : (output) Array of length N containing the abscissas
	COF : (input) Array of length 3*N containing the coefficients of
		the recurrence relation for orthogonal polynomials
		P_i(x)=(COF[3*i]*x+COF[3*i+1])P_{i-1}(x) - COF[3*i+2]*P_{i-2}(x)
	RI0 : (input) The integral of weight function over the required interval.
		
	Error status is returned by the value of the function GAUSRC.
		0 value implies successful execution
		302 implies N<=0
		321 implies that some coefficient becomes imaginary
			during calculations.
		In both these cases calculations are abandoned.
		Other values may be set by TQL2

     Required functions : TQL2
*/
 
#include <math.h>
#include <stdlib.h>

int tql2(double *z, int n, int iz, double d[], double e[], double reps);

int gausrc(int n, double w[], double ab[], double *cof, double ri0)

{
	int i,j,ier;
	double r1,reps=1.0e-15;
	double *wk,*d,*e;

	if(n<=0) return 302;
	wk=(double *) calloc((size_t) (n*n), sizeof(double));
	d=(double *) calloc((size_t) n, sizeof(double));
	e=(double *) calloc((size_t) n, sizeof(double));

/*     Calculate the coefficients of symmetric tridiagonal matrix */
	for(i=0; i<n; ++i) {
		d[i]=-cof[3*i+1]/cof[3*i];
		if(i<n-1) {
			r1=cof[3*i+5]/(cof[3*i]*cof[3*i+3]);
			if(r1>=0.0) e[i+1]=sqrt(r1);
			else {free(e); free(d); free(wk); return 321;}
		}
		for(j=0; j<n; ++j) wk[j+i*n]=0.0;
		wk[i+i*n]=1.0;
	}

/*     Find eigenvalues and eigenvectors of the tridiagonal matrix */
	ier=tql2(wk,n,n,d,e,reps);
	if(ier>0) {free(e); free(d); free(wk); return ier;}

/*     Calculate the abscissas and weights */
	for(i=0; i<n; ++i) {
		ab[i]=d[i];
		w[i]=wk[i]*wk[i]*ri0;
	}
	free(e); free(d); free(wk);
	return 0;
}
